squad/core/tasks/notification.py

```diff
@@ -8,6 +10,28 @@
 
 
 @celery.task
+def maybe_notify_project_status(status_id):
+    projectstatus = ProjectStatus.objects.get(pk=status_id)
+    build = projectstatus.build
+    project = build.project
+
+    if project.wait_before_notification:
+        to_wait = project.wait_before_notification
+        time_passed = timezone.now() - build.datetime
+        if time_passed.seconds < to_wait:
+            # wait time did not pass yet; try again later
+            remaining_time = to_wait - time_passed.seconds + 1
+            maybe_notify_project_status.apply_async(
```

Comment:
Any concern or attention to concurrency risk here? What if this gets called many times and then you have many of these execute concurrently - could it result in duplicate emails? Is the check at line 30 thread safe?

Reply:
Both of the values in line 30 are cached in line 14. After digging a bit into code it turns out that the flags are only set after email is sent. This might take a moment. UpdateProjectStatus is called on each test job finishing. So a task is triggered at this point. We won't see concurrency issues on our instance as we're only running one worker now. But I think it's a valid comment and needs to be addressed. The problem will manifest with more than one worker. We'll have a few tasks trying to send_status_notification for the same ProjectStatus. They're all scheduled with 'countdown' firing around the same time.